<a href="https://colab.research.google.com/github/JuCaRiCo/202602-AnaliticaDescriptivaPredictiva/blob/main/Actividad_Calificada_07_Datos_Temporales_271760.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformacion de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programacion para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformacion de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** Juan Carlos Rivera Cordova | |
| **Matrícula** 271760 | |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu codigo en las celdas marcadas con `# Escribe tu codigo aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu codigo**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificacion de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesion: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecucion → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripcion |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzo el **domingo 10 de marzo**: a las 02:00 el reloj salto a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversion y validacion** | 20 | 18–20: conversion con formato explícito y verificacion de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; codigo comentado; conclusion con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusion. | 1–13: conversion sin verificar o validaciones incompletas; conclusion sin sustento. | 0: no se entrego o el codigo no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duracion y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusion argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entrego o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretacion fundamentada. | 18–22: componentes y agrupaciones correctos; interpretacion incompleta. | 1–17: agrupaciones con errores o sin interpretacion. | 0: no se entrego o es incorrecto. |
| **4. Dataset limpio y comparacion** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparacion entre periodos con filtros temporales correcta; registro de decision completo. | 14–17: dataset limpio correcto; comparacion o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificacion. | 0: no se entrego o es incorrecto. |
| **Reflexion final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relacion con los resultados. | 0: no se entrego. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del codigo cuentan para la calificacion **de esa actividad**, no solo para la reflexion final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificacion es 0. El descuento se aplica sobre la calificacion obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversion y validacion (20 puntos)

### 1.1 Conversion (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversion no genero `NaT`: compara el número de faltantes antes y después.

In [2]:
print('1. Tipo de Dato')
print('pickup:',taxis['pickup'].dtype)
print('dropoff:', taxis['dropoff'].dtype)


valoresFaltantes = taxis[['pickup','dropoff']].isna().sum()

print('-----------------------------------')
print('2. Conversion')
taxis['pickup'] = pd.to_datetime(taxis['pickup'], format='%Y-%m-%d %H:%M:%S')
taxis['dropoff'] = pd.to_datetime(taxis['dropoff'], format='%Y-%m-%d %H:%M:%S')
print('Tipo de Dato despues de Conversion:\n', taxis[['pickup', 'dropoff']].dtypes)

print('-----------------------------------')
print('3. Verificacion')
print('Valores Faltantes antes de conversion:\n', valoresFaltantes)
valoresFaltantesDespues = taxis[['pickup','dropoff']].isna().sum()
print('Valores Faltantes despues de conversion:\n', valoresFaltantesDespues)

1. Tipo de Dato
pickup: object
dropoff: object
-----------------------------------
2. Conversión
Tipo de Dato despues de Conversion:
 pickup     datetime64[ns]
dropoff    datetime64[ns]
dtype: object
-----------------------------------
3. Verificación
Valores Faltantes antes de conversion:
 pickup     0
dropoff    0
dtype: int64
Valores Faltantes despues de conversion:
 pickup     0
dropoff    0
dtype: int64


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [3]:
print('1. Fechas minimas y maximas:')
print('Fecha minima de pickup:', taxis['pickup'].min())
print('Fecha maxima de pickup:', taxis['pickup'].max())
print('Fecha minima de dropoff :', taxis['dropoff'].min())
print('Fecha maxima de dropoff :', taxis['dropoff'].max())
print('-----------------------------------')
print('2. Pickup Viajes fuera de Marzo 2029')
pickupEnMarzo = (taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3)
print('Numero de Viajes fuera de Marzo: ', taxis[~pickupEnMarzo].shape[0])
print('Detalle de Viajes fuera de Marzo:\n', taxis[~pickupEnMarzo])
print('-----------------------------------')
print('3. Viajes pickup en Marzo y dropoff fuera de Marzo')
dropoffFueraMarzo = ~((taxis['dropoff'].dt.year == 2019) & (taxis['dropoff'].dt.month == 3))
pickupMarzoDropoffFuera = taxis[pickupEnMarzo & dropoffFueraMarzo]
print('pickup en Marzo y dropoff fuera de Marzo:', pickupMarzoDropoffFuera.shape[0])
print('Detalle de Viajes pickup en Marzo y dropoff fuera de Marzo:\n', pickupMarzoDropoffFuera)



1. Fechas minimas y maximas:
Fecha minima de pickup: 2019-02-28 23:29:03
Fecha maxima de pickup: 2019-03-31 23:43:45
Fecha minima de dropoff : 2019-02-28 23:32:35
Fecha maxima de dropoff : 2019-04-01 00:13:58
-----------------------------------
2. Pickup Viajes fuera de Marzo 2029
Numero de Viajes fuera de Marzo:  1
Detalle de Viajes fuera de Marzo:
                   pickup             dropoff  passengers  distance  fare  tip  \
6203 2019-02-28 23:29:03 2019-02-28 23:32:35           1       0.9   5.0  0.0   

      tolls  total  color payment  pickup_zone                   dropoff_zone  \
6203    0.0    6.3  green    cash  Old Astoria  Long Island City/Queens Plaza   

     pickup_borough dropoff_borough  
6203         Queens          Queens  
-----------------------------------
3. Viajes pickup en Marzo y dropoff fuera de Marzo
pickup en Marzo y dropoff fuera de Marzo: 1
Detalle de Viajes pickup en Marzo y dropoff fuera de Marzo:
                  pickup             dropoff  passenge

### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [4]:
print('1. Viajes por hora en el 10 de marzo')
viajesMarzo = taxis[(taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3) & (taxis['pickup'].dt.day == 10)]
print('Viajes Marzo:\n', viajesMarzo)

viajesPorHoraMarzo = viajesMarzo['pickup'].dt.hour.value_counts().sort_index()
print('Viajes por hora:\n', viajesPorHoraMarzo)

viajes2am = viajesPorHoraMarzo.get(2, 0) != 0
print('-----------------------------------')
print('Hay viajes para las 2:00: ', viajes2am)
print('-----------------------------------')
print('2. Localiza pickup en la zona')
pickupTz = taxis['pickup'].dt.tz_localize('America/New_York', nonexistent='NaT')
natCountTz = pickupTz.isna().sum()
print('Cantidad de valores NaT generados tras tz_localize(\'America/New_York\'):', natCountTz)

1. Viajes por hora en el 10 de marzo
Viajes Marzo:
                   pickup             dropoff  passengers  distance  fare  \
3    2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70  27.0   
30   2019-03-10 15:14:20 2019-03-10 15:20:17           1      1.84   7.0   
122  2019-03-10 20:36:47 2019-03-10 20:47:33           2      2.00   9.5   
130  2019-03-10 00:20:33 2019-03-10 00:27:13           3      0.57   6.0   
194  2019-03-10 06:32:41 2019-03-10 06:36:23           1      0.66   4.5   
...                  ...                 ...         ...       ...   ...   
6242 2019-03-10 18:08:13 2019-03-10 18:15:05           1      1.43   6.5   
6324 2019-03-10 10:29:40 2019-03-10 10:41:35           1      2.55  11.0   
6386 2019-03-10 10:40:11 2019-03-10 10:49:35           1      2.39   9.5   
6393 2019-03-10 18:33:13 2019-03-10 19:32:38           1      0.00  71.2   
6417 2019-03-10 12:10:45 2019-03-10 12:43:05           2     11.17  35.0   

       tip  tolls  total   color   

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### 1.4 Conclusion (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**
1. No se encontraron problemas al convertir las fechas
2. Solo se encontro un registro donde el dropoff esta fuera del mes, un porcentaje de 0.015%
3. Existe un viaje fuera del mes de marzo, una discrpancia del 0.015%
4. Tal parece que las fechas ya se encuentran dentro de la hora local para Nueva York
5. No se localizaron registros para el dia 10 de marzo a las 2am, esto debido a que es el momento en que se hace el cambio de horario para (Daylight Saving Time)
_Escribe aquí tu respuesta._

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duracion de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duracion en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [6]:
# 1 Creacion de columna duration
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']

# 2 Creacion de Duracion_min
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds() / 60

# 3 mresumen estadistico
taxis['duracion_min'].describe()

,duracion_min
count,6433.000000
mean,14.349617
std,11.643892
min,0.000000
25%,6.500000
50%,10.900000
75%,18.516667
max,107.666667


### 2.2 Duraciones anomalas (6 pts)

Cuenta los viajes con:

- Duracion **negativa** (termina antes de iniciar).
- Duracion **cero**.
- Duracion **menor a 1 minuto** (incluye los de duracion cero).
- Duracion **mayor a 3 horas**.

Muestra los viajes de duracion cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [30]:
# 1 Duracion Negativa
duracionNegativa = (taxis['pickup'] > taxis['dropoff']) | (taxis['duracion_min'] < 0)
print('Viajes con duracion negativa:', duracionNegativa.sum())

# 2 Duracion cero
duracionCero = taxis['duracion_min'] == 0
print('\nViajes con duracion cero:', duracionCero.sum())

# 3 Duracion \nmenor a 1 minuto
duracionMenor1Minuto = taxis['duracion_min'] < 1
print('\nViajes con duracion menor a 1 minuto:', duracionMenor1Minuto.sum())
print(taxis[taxis['duracion_min'] < 1])

# 4 Duracion mayor a 3 horas
duracionMayor3Horas = taxis['duracion_min'] > 180
print('\nViajes con duracion mayor a 3 horas:', duracionMayor3Horas.sum())

Viajes con duracion negativa: 0

Viajes con duracion cero: 6

Viajes con duracion menor a 1 minuto: 72
                  pickup             dropoff  passengers  distance  fare  \
42   2019-03-30 23:59:14 2019-03-30 23:59:17           1      0.00  80.0   
88   2019-03-16 05:08:02 2019-03-16 05:08:20           1      0.07   2.5   
120  2019-03-21 17:21:44 2019-03-21 17:21:49           1      0.00   2.5   
400  2019-03-09 17:26:02 2019-03-09 17:26:10           1      0.90   2.5   
458  2019-03-24 01:08:20 2019-03-24 01:08:34           1      0.02   2.5   
...                  ...                 ...         ...       ...   ...   
6332 2019-03-19 06:57:36 2019-03-19 06:58:15           1      0.08   2.5   
6355 2019-03-16 13:24:04 2019-03-16 13:24:13           1      0.00   2.5   
6356 2019-03-01 16:58:23 2019-03-01 16:58:23           1      0.00   3.0   
6357 2019-03-19 10:43:58 2019-03-19 10:44:07           1      0.00   2.5   
6372 2019-03-20 14:24:28 2019-03-20 14:24:32           1     


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duracion mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condicion)`, que deja `NaN` donde la condicion es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [34]:
# 1 Calcular velocidad y crear columna
convercionDeVelocidad = taxis['distance'] / (taxis['duracion_min'] / 60)

taxis['velocidad_mph'] = convercionDeVelocidad.where(taxis['duracion_min'] > 0)

# 2 contar viajes mayores a 60 mph
viajesMayores60mph = taxis['velocidad_mph'] > 60
print('Viajes con velocidad mayor a 60 mph:', viajesMayores60mph.sum())
print(taxis[taxis['velocidad_mph'] > 60])

# 3 Mostrar 5 viajes con mayor velocidad
colsVelocidad = ['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']
print('\n5 Viajes con mayor velocidad:')
print(taxis.sort_values(by='velocidad_mph', ascending=False)[colsVelocidad].head(5))

Viajes con velocidad mayor a 60 mph: 11
                  pickup             dropoff  passengers  distance   fare  \
400  2019-03-09 17:26:02 2019-03-09 17:26:10           1      0.90   2.50   
491  2019-03-07 07:11:33 2019-03-07 07:11:39           1      1.60   2.50   
770  2019-03-02 03:16:59 2019-03-02 03:17:06           0      9.40   2.50   
2387 2019-03-28 15:58:52 2019-03-28 15:59:25           1      1.80  69.06   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29           1      0.34  65.00   
3336 2019-03-06 09:24:48 2019-03-06 09:25:07           1      1.70  14.00   
3571 2019-03-22 16:47:41 2019-03-22 16:47:50           1      0.81  66.00   
4746 2019-03-12 07:06:07 2019-03-12 07:06:21           1      1.30   2.50   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11           1      1.70   2.50   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52           1     10.10  12.00   
6169 2019-03-27 02:11:01 2019-03-27 02:12:03           1      4.10   3.00   

        tip  tolls  total   color  

### 2.4 Criterio de exclusion (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duracion, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**

RESPUESTA A)
No todos los viajes menores son errores, dado que la tarifa es baja, posible cancelacion despues de iniciado, pero hay algunos que la tarifa es alta con 0 de distancia, en esos casos son errores

RESPUESTA B)
- Excluir viajes donde pickup sea fuera de marzo, ya que estan fuera del mes que se esta considerando
- Excluir viajes de duracion y distancia 0, ya que la finalidad de la movilidad depende del desplazamiento.
- Excluir los de velocidad mayor a 60 MPH que tienen una velocidad sumamente exagerada como por ejemplo 612 mph

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [37]:
# Creacion columna Hora
taxis['hora'] = taxis['pickup'].dt.hour
taxis['dia_num'] = taxis['pickup'].dt.dayofweek
taxis['dia_nombre'] = taxis['pickup'].dt.day_name()
taxis['fin_semana'] = taxis['dia_num'] >= 5
taxis['fecha'] = taxis['pickup'].dt.date
taxis['cruza_medianoche'] = taxis['fecha'] != taxis['fecha'].shift()
print('Viajes que cruzan la medianoche:', taxis['cruza_medianoche'].sum())
colsNuevas = ['hora', 'dia_num', 'dia_nombre', 'fin_semana', 'fecha', 'cruza_medianoche']
print('Nuevas Columnas:\n', taxis[colsNuevas].head(5))

Viajes que cruzan la medianoche: 6204
Nuevas Columnas:
    hora  dia_num dia_nombre  fin_semana       fecha  cruza_medianoche
0    20        5   Saturday        True  2019-03-23              True
1    16        0     Monday       False  2019-03-04              True
2    17        2  Wednesday       False  2019-03-27              True
3     1        6     Sunday        True  2019-03-10              True
4    13        5   Saturday        True  2019-03-30              True


### 3.2 Patron por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duracion tiene valores extremos.

In [40]:
# Agrupando y calculando
patronHora = taxis.groupby('hora').agg(
    num_viajes=('pickup', 'size'),
    mediana_duracion_min=('duracion_min', 'median')
)


# Identificar la hora con mayor y menor volumen de viajes
horaMasViajes = patronHora['num_viajes'].idxmax()
maxViajes = patronHora['num_viajes'].max()

horaMenosViajes = patronHora['num_viajes'].idxmin()
minViajes = patronHora['num_viajes'].min()

print('hora del día.:\n', patronHora)

print(f"\nHora con MÁS viajes:  {horaMasViajes}:00 hrs con {maxViajes} viajes.")
print(f"Hora con MENOS viajes: {horaMenosViajes}:00 hrs con {minViajes} viajes.")

hora del día.:
       num_viajes  mediana_duracion_min
hora                                  
0            205             10.733333
1            110              9.633333
2            101              9.183333
3             67              9.583333
4             57             10.500000
5             51              8.700000
6            142              7.833333
7            221              9.583333
8            313             11.316667
9            319             11.616667
10           325             11.883333
11           295             12.033333
12           334             12.775000
13           316             10.558333
14           360             11.841667
15           330             12.058333
16           336             11.583333
17           388             11.975000
18           417             11.100000
19           406             10.650000
20           368              9.258333
21           355             11.083333
22           321              9.816667
23       

### 3.3 Patron por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [44]:
# Agrupar por 'dia_num' y 'dia_nombre'
# Calculos: total de viajes, días unicos en el mes, mediana de duracion y mediana de tarifa (fare)
patronDia = taxis.groupby(['dia_num', 'dia_nombre']).agg(
    total_viajes=('pickup', 'size'),
    dias_unicos=('fecha', 'nunique'),
    mediana_duracion_min=('duracion_min', 'median'),
    mediana_fare=('fare', 'median')
)

# Calculo del promedio de viajes diarios para cada dia de la semana
patronDia['promedio_viajes_diario'] = patronDia['total_viajes'] / patronDia['dias_unicos']

print("Viajes por día de la semana (Lunes a Domingo)")
patronDia

Viajes por día de la semana (Lunes a Domingo)


,,total_viajes,dias_unicos,mediana_duracion_min,mediana_fare,promedio_viajes_diario
dia_num,dia_nombre,,,,,
0,Monday,708,4,10.541667,9.5,177.00
1,Tuesday,825,4,11.216667,9.5,206.25
2,Wednesday,966,4,11.291667,9.5,241.50
3,Thursday,905,5,11.766667,9.5,181.00
4,Friday,1115,5,11.150000,9.5,223.00
5,Saturday,1046,5,10.466667,9.0,209.20
6,Sunday,868,5,9.983333,9.0,173.60


### 3.4 Interpretacion (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duracion mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

RESPUESTA A)
- Por Hora: si asumimos que el horario productivo empieza a las 5am, la curva se muestra ascendente a partir de ahi, teniendo el pico mas alto a las 6pm
- Coincidencia de hora con mas viajes y hora mayor duracion mediana: el pico mas alto que es a las 6pm, hay 417 viajes, con una mediana de 11.10, y la hora con una mediana mas alta es a las 12pm con una mediana de 12.775000, y un  numero de viajes de 334, por lo cual si bien es alto, no coincide mucho con 417 viajes de las 6pm. El que si tiene una mayor conincidencia con las 6pm, es el de las 7pm con 406 viajes y una mediana de 10.650000, que no esta muy alejada con la de las 6pm

RESPUESTA B)
Ese mes inicio en viernes, y termino en domingo, lo cual hace que los viajes se carguen mas en viernes-domingo, lo cual nos podria dar un dato erroneo con respeccto a los demas dias, ya que se contabiliza un dia mas en fin de semana que en los dias entre semana.

---
## Actividad 4. Dataset limpio y comparacion (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condicion y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [47]:
# Criterio 1: Excluir pickups fuera de marzo de 2019
fueraMarzo = ~((taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3))

# Criterio 2: Excluir viajes sin recorrido físico o sin tiempo (distancia = 0 O duracion = 0)
sinTrayecto = (taxis['distance'] == 0) | (taxis['duracion_min'] == 0)

# Criterio 3: Excluir velocidades inviables en entorno urbano (> 80 mph)
velocidadExtrema = taxis['velocidad_mph'] > 80

# Contar viajes excluidos por cada criterio
excluidosMarzo = fueraMarzo.sum()
excluidosTrayecto = sinTrayecto.sum()
excluidosVelocidad = velocidadExtrema.sum()

# Combinar las mascara para obtener la mascara global de exclusion
excluirTotal = fueraMarzo | sinTrayecto | velocidadExtrema
totalExcluidos = excluirTotal.sum()

# Crear el DataFrame 'taxis_limpio' aplicando la inversion de la mascara de exclusion (~mask)
taxis_limpio = taxis[~excluirTotal].copy()

# Imprimir de resultados de la limpieza
print(f"Viajes fuera de marzo 2019:                {excluidosMarzo}")
print(f"Viajes sin trayecto (distancia 0 o dur 0):  {excluidosVelocidad}")
print(f"Viajes con velocidad > 80 mph:             {excluidosVelocidad}")
print(f"TOTAL de viajes únicos excluidos:          {totalExcluidos}")
print(f"Registros originales:                     {len(taxis)}")
print(f"Registros en 'taxis_limpio':              {len(taxis_limpio)}")

# 5. Verificaciones finales sobre 'taxis_limpio'
checkFueraMarzo = (~((taxis_limpio['pickup'].dt.year == 2019) & (taxis_limpio['pickup'].dt.month == 3))).sum()
checkDistCero = (taxis_limpio['distance'] == 0).sum()
checkDuracionCero = (taxis_limpio['duracion_min'] == 0).sum()
checkVelExtrema = (taxis_limpio['velocidad_mph'] > 80).sum()

print("\nVerificacion de 'taxis_limpio' ---")
print(f"Registros fuera de marzo 2019 restante:   {checkFueraMarzo}")
print(f"Registros con distancia = 0 restante:      {checkDistCero}")
print(f"Registros con duracion = 0 restante:       {checkDuracionCero}")
print(f"Registros con velocidad > 80 mph restante: {checkVelExtrema}")

Viajes fuera de marzo 2019:                1
Viajes sin trayecto (distancia 0 o dur 0):  11
Viajes con velocidad > 80 mph:             11
TOTAL de viajes únicos excluidos:          63
Registros originales:                     6433
Registros en 'taxis_limpio':              6370

Verificacion de 'taxis_limpio' ---
Registros fuera de marzo 2019 restante:   0
Registros con distancia = 0 restante:      0
Registros con duración = 0 restante:       0
Registros con velocidad > 80 mph restante: 0


### 4.2 Comparacion entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [49]:
# Se define la mascara para la franja horaria de la mañana (7:00 a 9:59 hrs)
manana = taxis_limpio['hora'].between(7, 9)

# Se Separa entre Dias Habiles (Lunes a Viernes: fin_semana == False) y Fin de Semana (Sabado y Domingo: fin_semana == True)
habiles = manana & (~taxis_limpio['fin_semana'])
finSemana = manana & (taxis_limpio['fin_semana'])

# Extraer subconjuntos de datos
viajesMananaHabiles = taxis_limpio[habiles]
viajesMananaFinSemana = taxis_limpio[finSemana]

# Construir la tabla comparativa con las metricas solicitadas
comparativa_manana = pd.DataFrame({
    'Días Habiles (Lun-Vie)': [
        len(viajesMananaHabiles),
        viajesMananaHabiles['duracion_min'].median(),
        viajesMananaHabiles['velocidad_mph'].median()
    ],
    'Fin de Semana (Sab-Dom)': [
        len(viajesMananaFinSemana),
        viajesMananaFinSemana['duracion_min'].median(),
        viajesMananaFinSemana['velocidad_mph'].median()
    ]
}, index=['Numero de viajes', 'Mediana duracion (min)', 'Mediana velocidad (mph)'])

print("Comparativa de viajes (07:00 a 09:59 hrs) ---")
print(comparativa_manana.round(2))

Comparativa de viajes (07:00 a 09:59 hrs) ---
                         Días Habiles (Lun-Vie)  Fin de Semana (Sab-Dom)
Numero de viajes                         684.00                   163.00
Mediana duracion (min)                    12.02                     7.63
Mediana velocidad (mph)                    9.11                    13.47


### 4.3 Registro de decision (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversion**: formato utilizado y verificacion.
- **Zona horaria**: en qué zona están los datos y como lo verificaste.
- **Criterio de exclusion**: condiciones, umbrales y justificacion.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

Documento de Registro de Decision de Datos (Data Cleaning Log)

Proyecto: Análisis de Movilidad y Patrones Temporales en Taxis de Nueva York

Dataset: taxis.csv (Marzo 2019)

Autor: Juan Carlos Rivera Cordova

Fecha de Elaboracion: Septiembre 2026

1. A) Conversion:
Se realizo la conversion explícita de las columnas de texto pickup y dropoff a objetos de tipo fecha y hora (datetime64[ns]) mediante la funcion pd.to_datetime() utilizando el formato ISO de representacion temporal: Formato: '\%Y-\%m-\%d \%H:\%M:\%S'

1. B) Antes de la conversion no habia valores nulos en pickup y dropoff. No habia valores invalidos


2. Zona Horaria. Los datos ya se econtraban en zona horaria local

3. Criterio de Exclusion:
- Excluir viajes donde pickup sea fuera de marzo, ya que estan fuera del mes que se esta considerando
- Excluir viajes de duracion y distancia 0, ya que la finalidad de la movilidad depende del desplazamiento.
- Excluir los de velocidad mayor a 60 MPH que tienen una velocidad sumamente exagerada como por ejemplo 612 mph
4. Hayazgos Principales: En la comparativa del horario matutino de las 07:00 a las 09:59 horas se observó que los días hábiles concentran una demanda casi cinco veces mayor con 853 viajes una duración mediana de 10.83 minutos y


---
## Reflexion final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situacion la inferencia habría producido errores silenciosos?
2. Supon que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Como lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duracion de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

1. Ventajas: Rendimiento y velocidad de computo, Robustez y Consistencia de los datos
Errores Silenciosos: estos se podrian presentar en las fechas que pudieran ser ambiguas, como 4 de marzo o 3 Abril.

2. De tener conocimiento de esta situacion se tendria que hacer el cambio para que ambas estuvieran en UTC o tiempo local

3. Habria un potencial sesgo por los registros de duracion cero y distancia cero, podria hacer que los calculos salieran mal. El otro riesgo es en las velocidades arriba de 60 mph, que estan desproporcionados

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesion.